internal/batches/executor/coordinator_test.go

```diff
@@ -379,12 +341,150 @@ func TestCoordinator_Execute(t *testing.T) {
 	}
 }
 
+func TestCoordinator_Execute_StepCaching(t *testing.T) {
+	cache := newInMemoryExecutionCache()
+
+	task := &Task{
+		Steps: []batches.Step{
+			{Run: `echo "one"`},
+			{Run: `echo "two"`},
+			{Run: `echo "three"`},
+		},
+		Repository:            testRepo1,
+		BatchChangeAttributes: &BatchChangeAttributes{},
+	}
+
+	executor := &dummyExecutor{}
+	executor.results = []taskResult{{
+		task: task,
+		result: executionResult{
+			Diff:         "dummydiff",
+			ChangedFiles: &git.Changes{},
+			Outputs:      map[string]interface{}{},
+			Path:         "",
+		},
+		stepResults: []stepExecutionResult{
+			{Step: 0, Diff: []byte(`step-0-diff`)},
+			{Step: 1, Diff: []byte(`step-1-diff`)},
+			{Step: 2, Diff: []byte(`step-2-diff`)},
+		},
+	}}
+
+	// First execution. Make sure that the Task executes all steps.
+	execAndEnsure(t, cache, executor, task, assertNoCachedResult(t))
+	// We now expect the cache to have 1+N entries: 1 for the complete task, N
+	// for the steps.
+	assertCacheSize(t, cache, len(task.Steps)+1)
```

Comment:
Ah, the spaces further down are there because of the parentheses I added (and I don't know why I added them). `go fmt` removes the spaces to show precedence: so `1 * 2 + 3` would be formatted as `1*2 + 3`. If I remove the parentheses, it gets formatted like you suggest.

Play around with the expressions here and hit the "Format" button at the top:

https://play.golang.org/p/2IY4f6ghQud

Reply:
Oooh tricky tricky. Makes sense though, thank you!